From: "Yan, Zheng" <zyan@redhat.com>
Date: Tue, 17 May 2016 11:58:02 +0800
Subject: ceph: call __fscache_uncache_page() if readpages fails
Git-commit: 368e35857dfab264f512d040c4486c9b13297988
Patch-mainline: v4.7-rc2
References: FATE#322288

If readpages fails, fscache needs to cleanup its internal state.

Signed-off-by: Yan, Zheng <zyan@redhat.com>
Acked-by: Luis Henriques <lhenriques@suse.com>
---
 fs/ceph/addr.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

--- a/fs/ceph/addr.c
+++ b/fs/ceph/addr.c
@@ -275,8 +275,10 @@ static void finish_read(struct ceph_osd_
 	for (i = 0; i < num_pages; i++) {
 		struct page *page = osd_data->pages[i];
 
-		if (rc < 0 && rc != -ENOENT)
+		if (rc < 0 && rc != -ENOENT) {
+			ceph_fscache_readpage_cancel(inode, page);
 			goto unlock;
+		}
 		if (bytes < (int)PAGE_CACHE_SIZE) {
 			/* zero (remainder of) page */
 			int s = bytes < 0 ? 0 : bytes;
